app/src/main/java/com/github/factotum_sdp/factotum/ui/signup/SignUpViewModel.kt

```diff
@@ -0,0 +1,32 @@
+package com.github.factotum_sdp.factotum.ui.signup
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.github.factotum_sdp.factotum.R
+import com.github.factotum_sdp.factotum.ui.auth.BaseAuthState
+
+class SignUpViewModel : ViewModel() {
+
+    private val _signupForm = MutableLiveData<SignUpFormState>()
+    val signupFormState: LiveData<SignUpFormState> = _signupForm
+
+    fun signUpDataChanged(username: String, email: String, password: String, role: String) {
+        if (!isUserNameValid(username)) {
+            _signupForm.value = SignUpFormState(usernameError = R.string.invalid_username)
+        } else if (!BaseAuthState.isEmailValid(email)) {
+            _signupForm.value = SignUpFormState(emailError = R.string.invalid_email)
+        } else if (!BaseAuthState.isPasswordValid(password)) {
+            _signupForm.value = SignUpFormState(passwordError = R.string.invalid_password)
+        } else if (role.isBlank()) {
+            //
+        } else {
+            _signupForm.value = SignUpFormState(isDataValid = true)
+        }
+    }
+
+    // A placeholder username validation check
+    private fun isUserNameValid(username: String): Boolean {
```

Comment:
Following the design of BaseAuthState, this method could be added to the companion object of the SignUpFormState class. But it's less important. I agree to merge and thanks for your changes.